crates/sp-domains/src/lib.rs

```diff
@@ -666,10 +682,39 @@ pub trait Domain {
     ) -> Option<H256> {
         use sp_externalities::ExternalitiesExt;
 
-        self.extension::<GenesisReceiptExtension>()
-            .expect("No `GenesisReceiptExtension` associated for the current context!")
+        self.extension::<GenesisStorageExtension>()
+            .expect("No `GenesisStorageExtension` associated for the current context!")
             .generate_genesis_state_root(domain_id, domain_instance_data)
     }
+
+    fn extract_runtime_info(&mut self, raw_genesis_storage: &[u8]) -> Option<RuntimeInfo> {
+        use sp_core::traits::ReadRuntimeVersionExt;
+        use sp_externalities::ExternalitiesExt;
+
+        let code = self
+            .extension::<GenesisStorageExtension>()
+            .expect("No `GenesisStorageExtension` associated for the current context!")
+            .extract_runtime_code(raw_genesis_storage)?;
+
+        let version = match self
+            .extension::<ReadRuntimeVersionExt>()
+            .expect("No `ReadRuntimeVersionExt` associated for the current context!")
+            .read_runtime_version(
+                code.as_slice(),
+                &mut sp_state_machine::BasicExternalities::default(),
+            )
+            .and_then(|v| RuntimeVersion::decode(&mut &v[..]).map_err(|e| e.to_string()))
+        {
+            Ok(v) => v,
+            Err(err) => {
+                log::warn!("cannot read version from the given runtime: {}", err);
+                return None;
+            }
+        };
+        let hash = BlakeTwo256::hash_of(&code);
```

Comment:
Why can't these be done in runtime without host function call? This looks like simple decoding of JSON and then getting values from well known keys to me unless I miss something. We don't need to work with "Storage" as such.